ast_tools/passes/cse.py

```diff
@@ -0,0 +1,197 @@
+        # also want to change behavior after recording the cse
+        if self.root is None:
+            self.root = node
+            return super().generic_visit(node)
+        elif self.recorded:
+            return self.replacer.visit(node)
+        else:
+            return super().visit(node)
+
+
+    def visit_Assign(self, node):
+        assert not self.recorded
+        finder = ExprFinder(self.cse)
+        finder.visit(node)
+        if finder.target is not None:
+            self.recorded = True
+            # eliminate the node from the expression
+            rest = self.replacer.visit(node)
+            save = ast.Assign(
+                        targets=[ast.Name(self.cse_name, ast.Store())],
+                        value=deepcopy(self.cse))
+            return [save, rest]
+        else:
+            return super().generic_visit(node)
+
+    # don't support control flow (assumes ssa)
+    def visit_If(self, node: ast.If):
+        raise SyntaxError(f"Cannot handle node {node}")
+
+    def visit_For(self, node: ast.For):
+        raise SyntaxError(f"Cannot handle node {node}")
+
+    def visit_AsyncFor(self, node: ast.AsyncFor):
+        raise SyntaxError(f"Cannot handle node {node}")
+
+    def visit_While(self, node: ast.While):
+        raise SyntaxError(f"Cannot handle node {node}")
+
+    def visit_With(self, node: ast.With):
+        raise SyntaxError(f"Cannot handle node {node}")
+
+    def visit_AsyncWith(self, node: ast.AsyncWith):
+        raise SyntaxError(f"Cannot handle node {node}")
+
+    def visit_Try(self, node: ast.Try):
+        raise SyntaxError(f"Cannot handle node {node}")
+
+    # don't recurs into defs
+    def visit_ClassDef(self, node: ast.ClassDef):
+        return node
+
+    def visit_FunctionDef(self, node: ast.FunctionDef):
+        return node
+
+    def visit_AsyncFunctionDef(self, node: ast.AsyncFunctionDef):
+        return node
+
+
+class cse(Pass):
+    '''
+    Performs common subexpression elimination
+
+    Most be run after ssa.
+
+    Post bool_to_bit will likely eliminate more as:
+        `a and b and c`
+    is a single `BoolOp`  but
+        `a & b & c`
+    is:
+        `(a & b) & c`
+
+    this means `a and b` is not a subexpression of `a and b and c`
+    but `a & b` is a subexpression of `a & b & c`
+    '''
+    def __init__(self,
+            cse_prefix: str = '__common_expr'):
+        self.cse_prefix = cse_prefix
+
+
+    def rewrite(self,
+            tree: ast.AST,
+            env: SymbolTable,
+            metadata: tp.MutableMapping) -> PASS_ARGS_T:
+
+        prefix = gen_free_prefix(tree, env, self.cse_prefix)
+        c = 0
+        while True:
+            counter = ExprCounter()
+            counter.visit(tree)
+            if not counter.cses:
+                break
+
+            expr, freq = counter.cses.most_common()[0]
+            if freq <= 1:
+                break
+
+            expr = mutable(expr)
+            elim = CSEElim(expr, prefix + repr(c))
```

Comment:
I was wondering why repr vs str, then found this https://stackoverflow.com/questions/45376719/why-is-reprint-faster-than-strint

Reply:
I didn't really think about it deeply.   I repr and str are the same for int so I just picked one.